Added missing class event handler methods

TomEdwardsEnscape opened the pull request:
Class event handlers methods are virtual, and so give derived classes full control over class event handling. Without them the only option would be to register a new class event handler, which would only execute after any base handler. A method gives the derived class the chance to execute code before the base class does, or not to call the base class method at all. I found some cases where methods which should exist did not.

* `InputElement.OnTapped`
* `InputElement.OnRightTapped`
* `InputElement.OnDoubleTapped`
* `InputElement.OnHolding`
* ~`MenuItem.OnPointerEnteredItem`~
* ~`MenuItem.OnPointerExitedItem`~

### Moved handling of `Holding` event in `Control` to the class method

This allows derived types to more easily override their holding behaviour, and eliminates one event subscription per `Control`, which is a very common type.

I didn't find any other instance subscriptions which could be converted to class handlers.

## Breaking changes
Because `Control` was subscribing to the `Holding` event in `OnAttachedToVisualTree`, there was an opportunity for an event subscription to be added before this point, which could then handle the hold event and prevent the built-in event handler from executing. This can be much more easily achieved per-instance by setting the attached property `Gesture.IsHoldingEnabled` to `false`. And of course, you can now disable/replace the behaviour entirely in a derived class.

## Obsoletions / Deprecations
None.

kekekeks commented:
Thank you for your pull request.


Our general policy is to not have two unrelated things in a single PR since:
- it's easier to review
- it's easier to pin down regressions via git bisect
- we sometimes have to revert changes and with separate PRs we can have more fine-grained control of what we merge and revert
- different changes have different backport/release policy

So, regarding the changes:

- Please, split pure documentation changes into a separate PR.
- GotFocus/LostFocus is a good catch, WPF invokes the event handler after setting any kind of private state too, so Avalonia should be matching that behavior.
- Missing class event handlers _might_ have been intentional since default handlers methods for events aren't exactly free, event route builder can skip calls if there are no handlers attached, it would need to be discussed separately too.
- OnHolding is, again, an unrelated change with it's own breaking behaviors, so unlike the focus thing we can't backport it

TomEdwardsEnscape committed:
Added missing class event handler methods

TomEdwardsEnscape committed:
Moved handling of Holding event in Control to the class method

TomEdwardsEnscape commented:
The new PRs are #18868 and #18869. I couldn't split the Holding change away because it depends on the previous commit. I'll just remove it from this PR if the final call is to backport them to different versions.

> Missing class event handlers might have been intentional since default handlers methods for events aren't exactly free, event route builder can skip calls if there are no handlers attached.

The events are raised at most once per frame per pointer. This is completely negligible.

TomEdwardsEnscape committed:
Exposed Gesture.RightTappedEvent as InputElement.RightTapped and added a class handler
This event was already being raised, but you wouldn't know about it unless you looked into Avalonia's internals

TomEdwardsEnscape commented:
> Could you please explain why you think those virtual methods should be added?

For consistency with `InputElement.OnPointerEntered` and `InputElement.OnPointerExited`.

I'm of the opinion that every class which defines a `RoutedEvent` should also define a virtual handler method for it, unless that would create a performance problem. But this PR isn't meant to exhaustively change everything to that pattern and I don't mind removing the `MenuItem` methods.

MrJul commented:
> Long answer: I'm of the opinion that every class which defines a `RoutedEvent` should also define a virtual handler method for it, unless that would create a performance problem. 

Thanks for sharing. The problem I personally have with that approach is that it isn't a performance problem until it is, and at that point it's too late...

Clicking? No problem, this is a low frequency operation, humans are slow by design. Moving the pointer? Well, that generates a lot of allocations today, but at the same time there's always some state to update when the pointer moves so it's never for nothing. Now any event that might get raised as part of the layout process? It's so easy to add thousands upon thousands of useless virtual calls that nobody ever overrides, impacting the layout times. Adding an event handler manually when needed should be a valid approach.

I'm mentioning this only for the sake of discussion :) Other team members might disagree with me. It's always a hard balance between performance and ease of use.

> But this PR isn't meant to exhaustively change everything to that pattern and I don't mind removing the `MenuItem` methods.

Please keep only the `InputElement` changes for this PR. We can have a more general discussion about the virtual handler methods.

MrJul reviewed (approved):
LGTM, thank you!
